logr.go

```diff
@@ -371,12 +375,49 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the logger. If the logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	l.ctx = ctx
+	return l
+}
+
+// WithContextValues stores a list of context keys and the name for them. When
+// given a context through WithContext later, the logger will extract the
+// values for these keys and log them as additional key/value pairs.
+func (l Logger) WithContextValues(keys ...ContextKey) Logger {
+	l.contextKeys = &keys
+	return l
+}
+
+func (l Logger) appendFromContext(keysAndValues []interface{}) []interface{} {
+	if l.ctx == nil || l.contextKeys == nil {
+		return keysAndValues
+	}
+	for _, key := range *l.contextKeys {
+		if value := l.ctx.Value(key.Key); value != nil {
+			keysAndValues = append(keysAndValues, key.Name, value)
+		}
+	}
+	return keysAndValues
+}
+
+// ContextKey defines the string that is to be used in a key/value pair when
+// logging the value that is stored in a context for the given key.
+type ContextKey struct {
+	Key  interface{}
+	Name string
+}
+
 // contextKey is how we find Loggers in a context.Context.
 type contextKey struct{}
 
 // FromContext returns a Logger from ctx or an error if no Logger is found.
+// WithContext gets called for the logger automatically.
```

Comment:
This sentence is unclear to me.  Maybe:

```
// FromContext returns a Logger from ctx or an error if no Logger is found.
// The value returned has already had WithContext called.
```

Same below

Reply:
Changed.